unix: machine_mem improvements

dhylands opened the pull request:
This basically introduces the MICROPY_MACHINE_MEM_GET_READ_ADDR
and MICROPY_MACHINE_MEM_GET_WRITE_ADDR macros. If one of them is
not defined, then a default identity function is provided.

dhylands committed:
unix: machine_mem improvements

This basically introduces the MICROPY_MACHINE_MEM_GET_READ_ADDR
and MICROPY_MACHINE_MEM_GET_WRITE_ADDR macros. If one of them is
not defined, then a default identity function is provided.

pfalcon commented:
From 1st look, this looks a bit overengineered. Half-look more, it seems like this might have been done to avoid need for duplicate read/write func. There's still opportunity to simplify/cleanup. Will add comments inline.

pfalcon commented on extmod/machine_mem.h:
```diff
@@ -41,10 +41,28 @@ extern const machine_mem_obj_t machine_mem8_obj;
+// function and can be found in machine_mem.c
 
-#endif /*  MICROPY_EXTMOD_MACHINE_MEM */
+#if !defined(MICROPY_MACHINE_MEM_GET_READ_ADDR)
```
This might have been put here to complement comment above, but doesn't really belong here - this logic can happen in .c, without macro creep (and especially without naming internal macros as if they were public, but again, doing it in .c, will save from need to have such macros at all).

~~~
#ifndef MICROPY_MACHINE_MEM_GET_WRITE_ADDR
int f() {}
#define MICROPY_MACHINE_MEM_GET_WRITE_ADDR foo
#endif
~~~

dhylands commented:
Yeah. This was in response to you wanting to add the identity function and not use weak functions.

dhylands commented:
moved some of #if stuff into the C file. Reworked @pfalcon's suggested way of doing it so that I still only wind up with one identity function rather than 2 if neither read nor write is overridden.

pfalcon commented:
Merged, thanks.
